web_poet/serialization/functions.py

```diff
@@ -227,3 +230,63 @@ def _deserialize_AnnotatedInstance(
 
 
 register_serialization(_serialize_AnnotatedInstance, _deserialize_AnnotatedInstance)
+
+
+def _serialize_BrowserHtml(o: BrowserHtml) -> SerializedLeafData:
+    return {"body.html": o.encode("utf8")}
+
+
+def _deserialize_BrowserHtml(
+    cls: type[BrowserHtml], data: SerializedLeafData
+) -> BrowserHtml:
+    body: bytes = data.get("body.html") or b""
+    return cls(body.decode("utf8"))
+
+
+register_serialization(_serialize_BrowserHtml, _deserialize_BrowserHtml)
+
+
+def _serialize_BrowserResponse(o: BrowserResponse) -> SerializedLeafData:
+    info = {
+        "url": str(o.url),
+        "status": o.status,
+    }
+    return {
+        "body.html": o.html.encode("utf8"),
+        "info.json": _format_json(info).encode(),
+    }
+
+
+def _deserialize_BrowserResponse(
+    cls: type[BrowserResponse], data: SerializedLeafData
+) -> BrowserResponse:
+    html = BrowserHtml(data["body.html"].decode("utf8"))
+    info = json.loads(data["info.json"])
+    return cls(
+        url=info["url"],
+        html=html,
+        status=info["status"],
+    )
+
+
+register_serialization(_serialize_BrowserResponse, _deserialize_BrowserResponse)
+
+
+def _serialize_AnyResponse(o: AnyResponse) -> SerializedLeafData:
+    if isinstance(o.response, HttpResponse):
+        return _serialize_HttpResponse(o.response)
+    return _serialize_BrowserResponse(o.response)
+
+
+def _deserialize_AnyResponse(
+    cls: type[AnyResponse], data: SerializedLeafData
+) -> AnyResponse:
+    info = json.loads(data["info.json"])
+    if "_encoding" in info:
```

Comment:
That's a nice trick, but I wonder if we can make it more explicit - e.g. set the type name (HttpResponse / BrowserResponse) in info.json, both for browser and non-browser requests. Then, you can read it and check for BrowserResponse (not HttpResponse, as past fixtures won't have it set).

Reply:
> not HttpResponse, as past fixtures won't have it set

I believe this includes not assuming that "type" is present in the loaded dict as well.